core/runtime/src/main/java/io/quarkus/runtime/LocalesBuildTimeConfig.java

```diff
@@ -44,7 +44,8 @@ public class LocalesBuildTimeConfig {
      * For instance, the Hibernate Validator extension makes use of it.
      * <p>
      * Native-image build uses this property to derive {@code user.language} and {@code user.country} for the application's
-     * runtime.
+     * runtime. Starting with GraalVM for JDK 24 this option will not result in setting the default runtime locale, and
+     * any usages of it should be re-evaluated.
      */
     @ConfigItem(defaultValue = DEFAULT_LANGUAGE + "-" + DEFAULT_COUNTRY, defaultValueDocumentation = "Build system locale")
```

Comment:
> * System default locale at image runtime with GraalVM for JDK 24 and better

Note that this is not enforced though, i.e., if we compile with `-Dquarkus.default-locale=cs-CZ` we will still have to pass `-Duser.language=cs` and/or `-Duser.country=CZ` at runtime. Note that passing `-Dquarkus.default-locale=cs-CZ` won't do the trick as it's considered build-time fixed, which I guess I need to fix as well.

I added a new test for this ^

Reply:
@gsmet any suggestions on how to approach this? To my understanding `quarkus.default-locale` should move to `RUN_TIME` config phase starting with GraalVM for JDK 24.

Reply:
Hum, I'm not sure we have a good way to achieve that. We moved some build time properties to runtime but it was an all or nothing thing.

I think we will have to keep it build time until we officially switch to a GraalVM/Mandrel version that is 24+.

Except if @radcortez comes up with some crazy ideas but I'm not entirely sure it's worth the hassle if we can live with it for now.

Reply:
Mappings allow you to map the same property for both build and runtime. It can either be done by another mapping, or just by registering the `LocalesBuildTimeConfig` in the runtime config customizer. 

We had a similar case with logging:
https://github.com/quarkusio/quarkus/pull/42157

Reply:
> > * System default locale at image runtime with GraalVM for JDK 24 and better
> 
> Note that this is not enforced though, i.e., if we compile with `-Dquarkus.default-locale=cs-CZ` we will still have to pass `-Duser.language=cs` and/or `-Duser.country=CZ` at runtime. Note that passing `-Dquarkus.default-locale=cs-CZ` won't do the trick as it's considered build-time fixed, which I guess I need to fix as well.
> 
> I added a new test for this ^

I'm not sure I follow. What I meant is that the native image follows the system locale at runtime. Yeah it's no longer build-time static (the default locale). For latest GraalVM master I see this. Test with `Locale.getDefault()` in a native image.

On `de-AT` locale:
```
$ localectl status
System Locale: LANG=de_AT.UTF-8
    VC Keymap: us
   X11 Layout: us
    X11 Model: pc105
$ ./locale-test-24 
Default locale: de_AT
```

On `en-US` locale:
```
$ localectl status
System Locale: LANG=en_US.UTF-8
    VC Keymap: us
   X11 Layout: us
$ ./locale-test-24 
Default locale: en_US
```

So with GraalVM for JDK 24, native images behave more like `java` would.

Reply:
Exactly, which means that `defaultLocale` currently has no effect at the native-image starting with GraalVM for JDK 24 (other than including the locale in the image). I am starting to believe that the best thing to do is to print a warning saying that `defaultLocale` is ignored (when using GraalVM for JDK 24) and deprecate it once GraalVM for JDK 25 is out.

I think adding the default locale to the list of included locales is still a good thing to do.

Reply:
@gsmet @Karm WDYT about:

> I am starting to believe that the best thing to do is to print a warning saying that `defaultLocale` is ignored (when using GraalVM for JDK 24) and deprecate it once GraalVM for JDK 25 is out.

Reply:
@radcortez I made `defaultLocale` an `Optional` and thus removed the `defaultValue` from its annotation, but that resulted in the following exception:

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Set.contains(Object)" because "this.val$localesBuildTimeConfig.locales" is null
	at io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder$2.created(HibernateValidatorRecorder.java:87)
	at io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:80)
	at io.quarkus.deployment.steps.ArcProcessor$notifyBeanContainerListeners1304312071.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ArcProcessor$notifyBeanContainerListeners1304312071.deploy(Unknown Source)
	... 42 more
```

See https://github.com/zakkak/quarkus/actions/runs/11203768674/job/31141630432#step:16:636

Note that the field it's complaining about is the `locales` one (which I didn't change and I kept its `defaultValue`) and not the `defaultLocale`.

Is that expected behavior? Am I missing some configuration?